### PR TITLE
[release/2.3] Enable bf16 with fp32 weights for MIOpen batchnorm

### PR DESCRIPTION
**Pull request by jithunnair-amd**
This PR enables:
* using MIOpen OCL_mix backend for bf16 batchnorm with fp32 weights (using torch autocast). This was required and tested for customer workload using NCHW (which is the only memory_layout enabled).
* logging for MIOpen batchnorm using `PYTORCH_MIOPEN_EXTRA_LOGGING` env var.

TODO in separate PR: Need to implement PyTorch unit tests for this bf16/fp16 inputs + fp32 weights case.

### CODE REVIEW

**Comment by rocm-mici** [comment]
Jenkins build for [a4303a659a1807cb764bc96f1f31b39031d1cd15](https://github.com/ROCm/pytorch/commit/a4303a659a1807cb764bc96f1f31b39031d1cd15) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1666/2//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1666/2/artifact)

**Comment by rocm-mici** [comment]
Jenkins build for [2ec5a3cd6f2410c5e9109266935368c2e97bce32](https://github.com/ROCm/pytorch/commit/2ec5a3cd6f2410c5e9109266935368c2e97bce32) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1666/3//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1666/3/artifact)

**Comment by rocm-mici** [comment]
Jenkins build for [3a0bee39bb7c9cc46be2f7719b96910e77c6a994](https://github.com/ROCm/pytorch/commit/3a0bee39bb7c9cc46be2f7719b96910e77c6a994) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1666/4//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1666/4/artifact)

**Review by pruthvistony — APPROVED** [review]
LGTM. Some cleanup is required, but good for initial implementation push.

**Comment by rocm-mici** [comment]
Jenkins build for [3a0bee39bb7c9cc46be2f7719b96910e77c6a994](https://github.com/ROCm/pytorch/commit/3a0bee39bb7c9cc46be2f7719b96910e77c6a994) commit finished as FAILURE
Links: [Blue Ocean view](http://ml-ci-internal.amd.com:8080/blue/organizations/jenkins/pytorch%2Fpytorch-ci-pipeline/detail/PR-1666/5//pipeline) / [Build artifacts](http://ml-ci-internal.amd.com:8080/job/pytorch/job/pytorch-ci-pipeline/job/PR-1666/5/artifact)

**Comment by jithunnair-amd** [comment]
!cherry-pick --onto release/2.4
